### PR TITLE
feat(tree2): Add initial infrastructure for Map proxies

### PR DESCRIPTION
**Pull request by Josmithr**
Currently allows creation of trees using Maps, and allows reading of their data. Mutation APIs are not possible yet, as they are not implemented on the underlying `MapNode` interface.

### CODE REVIEW

File: experimental/dds/tree2/src/feature-libraries/contextuallyTyped.ts [hunk 1]
```diff
@@ -552,7 +552,27 @@ export function applyTypesFromContext(
 			type,
 			fields: new Map(children.length > 0 ? [[primary.key, children]] : []),
 		};
+	} else if (data instanceof Map) {
+		// Map
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/contextuallyTyped.ts:556]: TODO: expand on the comment here to actually explain what this path is doing.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts [hunk 2]
```diff
@@ -497,3 +514,94 @@ export function createListProxy<TTypes extends AllowedTypes>(
...
+	Map.prototype.entries,
+	Map.prototype.forEach,
+	Map.prototype.get,
+	Map.prototype.has,
+	Map.prototype.keys,
+	Map.prototype.set,
+	Map.prototype.values,
+	/* eslint-enable @typescript-eslint/unbound-method */
+].forEach((fn) => {
+	listPrototypeProperties[fn.name] = { value: fn };
```
Comment (DLehenbauer) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts:535]: Whoops.  Don't add them to list's prototype.  :-)

Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts:535]: Oof, that was definitely bound to the correct variable earlier. I don't know how I messed that up in the merge.

Okay, I was wondering if that might be the case. I'd discovered that get didn't work as expected, but I was hoping some of the other methods might "just work". If that's not the case, I'll go ahead and just clear this out altogether and leave placeholders in the function.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts [hunk 3]
```diff
@@ -497,3 +514,94 @@ export function createListProxy<TTypes extends AllowedTypes>(
...
+	return new Proxy<SharedTreeMap<TSchema>>(new Map<string, ProxyNode<TSchema>>(), {
+		get: (target, key, receiver): unknown => {
+			console.log(`get "${String(key)}"`);
+
+			if (key === "get") {
+				return (mapKey: string) => {
+					const mapNode = getMapNode(dispatch);
+					return getProxyForField(mapNode.getBoxed(mapKey));
+				};
+			}
```
Comment (DLehenbauer) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts:588]: Here you're intercepting "get" and returning your own implementation before looking it up in the dispatch object.

Just move this implementation to the dispatch object's prototype, per the above.

We're essentially using the dispatch object (and prototype) to replace this switch:

switch(key) {
    case "get": ...
    case "set": ...
    case "has": ...
    ...
}

With a "map" of string -> Function.

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/proxies.spec.ts [hunk 4]
```diff
@@ -90,6 +103,13 @@ describe("SharedTreeObject", () => {
 		}
 	});
 
+	itWithRoot("can read maps", schema, initialTree, (root) => {
+		assert.equal(root.map.size, 2);
+		assert.equal(root.map.get("foo"), "Hello");
+		assert.equal(root.map.get("bar"), "World");
+		assert.equal(root.map.get("baz"), undefined);
+	});
```
Comment (DLehenbauer) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/proxies.spec.ts:111]: Nice!

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts (hunk 3 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts:572]: Ah, okay. Makes sense. Will update!

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts [hunk 5]
```diff
@@ -497,3 +514,76 @@ export function createListProxy<TTypes extends AllowedTypes>(
...
+		has: (target, key) => {
+			return Reflect.has(dispatch, key);
+		},
+		set: (target, key, newValue): boolean => {
+			// For MVP, we otherwise disallow setting properties (mutation is only available via the map mutation APIs).
+			return false;
+		},
+		ownKeys: (target) => {
+			// TODO: double check that this is all we require here
+			return ["size"];
```
Comment (DLehenbauer) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts:586]: Interesting.  On Map, all properties (inculding 'size') are inherited via the prototype:

```
> Reflect.ownKeys(new Map([[0,1]]))
[]
```

So 'ownKeys' should return [], getOwnPropertyDescriptor always returns undefined, etc.

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/proxies.spec.ts [hunk 6]
```diff
@@ -90,6 +103,32 @@ describe("SharedTreeObject", () => {
...
+		const result = [];
+		for (const entry of root.map) {
+			result.push(entry);
+		}
+
+		assert.deepEqual(result, [
+			["foo", "Hello"],
+			["bar", "World"],
+		]);
+	});
```
Comment (DLehenbauer) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree-2/proxies.spec.ts:130]: Looking good!

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts (hunk 5 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts:586]: Ah, I see. And thanks for the tip on how to explicitly verify! I'll get this updated and add some comments!

File: experimental/dds/tree2/src/feature-libraries/contextuallyTyped.ts [hunk 7]
```diff
@@ -552,6 +552,24 @@ export function applyTypesFromContext(
 			type,
 			fields: new Map(children.length > 0 ? [[primary.key, children]] : []),
 		};
+	} else if (data instanceof Map) {
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/contextuallyTyped.ts:555]: @CraigMacomber hoping you can provide feedback / guidance here. Noah and I thought this was likely the best place to handle JS Map inputs. The non-proxy API won't allow such inputs per the type-system, but if we ever allowed that, it should presumably be fine? And for the proxy API, this enables Map inputs without having to re-implement the rest of the recursive logic here.

Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/contextuallyTyped.ts:555]: this entire file is logic that should be superseded by Noah's factory stuff. Also this is just intended for making nodes out of object literals with strong types from  SchemaAware's flexiable mode. Adding support for maps here doesn't seem very useful (they are already supported as object literals, which is the intended use of this API) and programmatic non schema aware creators of trees shouldn't use this contextually typed format (they should make something else, like MapTree, then wrap it in a cursor).

If you do want to add this here (and I don't think its worth doing) you would also want to update the schema aware APi we use with it so users (of the old editiable tree API and initlTree until we change it) could actually benefit from your feature.

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts [hunk 8]
```diff
@@ -497,3 +514,90 @@ export function createListProxy<TTypes extends AllowedTypes>(
...
+
+const mapStaticDispatchMap: PropertyDescriptorMap = {
+	[Symbol.iterator]: {
+		value(this: SharedTreeMap<MapSchema>) {
+			const node = getMapNode(this);
+			return node[Symbol.iterator]();
+		},
+	},
+	entries: {
+		value(): IterableIterator<[string, unknown]> {
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts:523]: TODO: Add `this` parameter?

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts [hunk 9]
```diff
@@ -497,3 +514,90 @@ export function createListProxy<TTypes extends AllowedTypes>(
...
+// #endregion
+
+const getMapNode = <TSchema extends MapSchema>(target: object): MapNode<TSchema> => {
+	return getTreeNode(target) as MapNode<TSchema>;
+};
+
+function createMapProxy<TSchema extends MapSchema>(treeNode: TreeNode): SharedTreeMap<TSchema> {
+	// Create a 'dispatch' object that this Proxy forwards to instead of the proxy target.
+	// Own properties on the dispatch object are surfaced as own properties of the proxy.
+	// (e.g., 'size', which is defined below).
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts:572]:
```suggestion
```

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts [hunk 10]
```diff
@@ -497,3 +514,90 @@ export function createListProxy<TTypes extends AllowedTypes>(
...
+	const dispatch = Object.create(mapPrototype, {});
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts:573]: Add a comment

File: experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts [hunk 11]
```diff
@@ -497,3 +514,90 @@ export function createListProxy<TTypes extends AllowedTypes>(
...
+			return Reflect.get(dispatch, key);
+		},
+		getOwnPropertyDescriptor: (target, key): PropertyDescriptor | undefined => {
+			return Reflect.getOwnPropertyDescriptor(dispatch, key);
+		},
+		has: (target, key) => {
+			return Reflect.has(dispatch, key);
+		},
+		set: (target, key, newValue): boolean => {
+			// For MVP, we otherwise disallow setting properties (mutation is only available via the map mutation APIs).
```
Comment (Josmithr) [inline experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxies.ts:590]: Update comment in context of Map

**Review by noencke — APPROVED** [review]
Reviewed online together - looks good!

File: experimental/dds/tree2/src/feature-libraries/contextuallyTyped.ts (hunk 7 above)
Comment (Josmithr) [reply experimental/dds/tree2/src/feature-libraries/contextuallyTyped.ts:555]: Per our offline conversation, we've agreed that this isn't the correct long-term home for this logic, but will suffice as a short term solution.

AB#6059 is tracking the longer term solution.

**Pull request merged**